### PR TITLE
feat: Add the Hover Range capability which enables showing the type of an expression

### PR DESCRIPTION
**Pull request by alexfertel**
Closes https://github.com/rust-analyzer/rust-analyzer/issues/389

This PR extends the `textDocument/hover` method to allow getting the type of an expression. It looks like this:

type_of_expression

Edit: One thing I noticed is that when hovering a selection that includes a macro it doesn't work, so maybe this would need a follow-up issue discussing what problem that may have.

(PS: What a great project! I am learning a lot! 🚀)

### CODE REVIEW

**Comment by matklad** [comment]
Oh my, I didn't realise that it's possible to convince VS Code to actually use this! My original plan was to just implement this in the server and wait for emacs/vim to pick this up, and then bug vs code maintainers to follow the suite. You, however, managed to implement this in vscode, kudos!

**Comment by matklad** [comment]
Note that we already have a custom hover request (we customize the return type), so you only need to add custom params:

```diff
diff --git a/crates/rust-analyzer/src/handlers.rs b/crates/rust-analyzer/src/handlers.rs
index f89b7f5d7..8a115ec30 100644
--- a/crates/rust-analyzer/src/handlers.rs
+++ b/crates/rust-analyzer/src/handlers.rs
@@ -867,7 +867,7 @@ pub(crate) fn handle_signature_help(
 
 pub(crate) fn handle_hover(
     snap: GlobalStateSnapshot,
-    params: lsp_types::HoverParams,
+    params: lsp_ext::HoverParams,
 ) -> Result<Option<lsp_ext::Hover>> {
     let _p = profile::span("handle_hover");
     let position = from_proto::file_position(&snap, params.text_document_position_params)?;
diff --git a/crates/rust-analyzer/src/lsp_ext.rs b/crates/rust-analyzer/src/lsp_ext.rs
index 7aed93f99..627cacc9c 100644
--- a/crates/rust-analyzer/src/lsp_ext.rs
+++ b/crates/rust-analyzer/src/lsp_ext.rs
@@ -376,24 +376,25 @@ pub struct SnippetTextEdit {
 pub enum HoverRequest {}
 
 impl Request for HoverRequest {
-    type Params = lsp_types::HoverParams;
+    type Params = HoverParams;
     type Result = Option<Hover>;
     const METHOD: &'static str = "textDocument/hover";
 }
 
-pub enum HoverRangeRequest {}
-
-impl Request for HoverRangeRequest {
-    type Params = HoverRangeParams;
-    type Result = Option<Hover>;
-    const METHOD: &'static str = "rust-analyzer/hoverRange";
-}
-
-#[derive(Deserialize, Serialize, Debug)]
+#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
 #[serde(rename_all = "camelCase")]
-pub struct HoverRangeParams {
+pub struct HoverParams {
     pub text_document: TextDocumentIdentifier,
-    pub range: Range,
+    pub position: PositionOrRange,
+    #[serde(flatten)]
+    pub work_done_progress_params: WorkDoneProgressParams,
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(untagged)]
+enum PositionOrRange {
+    Position(lsp_types::Position),
+    Range(lsp_types::Range),
 }
 
 #[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
```

**Comment by matklad** [comment]
And don't forget to document the extension in the `lsp-extensions.md` and add a corresponding server capability for it.

**Comment by alexfertel** [comment]
> Note that we already have a custom hover request (we customize the return type), so you only need to add custom params:

This is it!!! My god, I struggled so much because I thought the name of the param had to change 😭

Thanks for the quick response! I'll try to revisit this today 😁

**Comment by flodiebold** [comment]
If it's just an extension to the existing hover request, I think this should be very easy to add in Emacs as well.

**Comment by alexfertel** [comment]
I think this is fairly complete :) No clue on how to approach adding the client/server capability 😅 Halp!

File: crates/ide/src/hover.rs
Comment (matklad) [inline crates/ide/src/hover.rs:79]: This needs to use only two slashes, otherwise it won't be picked up by our docgen infra.

File: crates/ide/src/lib.rs
Comment (matklad) [inline crates/ide/src/lib.rs:427]: Rather than adding a second function, let's change the existing one to take a range, and let's use empty range for the case where the input is a single position.

File: docs/dev/lsp-extensions.md
Comment (matklad) [inline docs/dev/lsp-extensions.md:661]: This should have `**Experimental Server Capability:** { "hoverRange": boolean }`, see how thouse are defiened for some other requests in this file.

Comment (matklad) [reply docs/dev/lsp-extensions.md:661]: You'll want to add an actual capability to https://github.com/rust-analyzer/rust-analyzer/blob/3a59b56324db99475ea2027cb94f82785b16ad14/crates/rust-analyzer/src/caps.rs#L115-L125

File: crates/ide/src/hover.rs
Comment (lnicola) [inline crates/ide/src/hover.rs:79]:
```suggestion
// image::https://user-images.githubusercontent.com/48062697/113020658-b5f98b80-917a-11eb-9f88-3dbc27320c95.gif[]
```

**Review by lnicola — APPROVED** [review]

**Review by matklad — APPROVED** [review]

**Comment by lnicola** [comment]
Not exactly, but...

changelog feat (first contribution) add "Hover Range" capability to get the type of an expression

File: crates/rust-analyzer/src/lsp_ext.rs
Comment (matklad) [inline crates/rust-analyzer/src/lsp_ext.rs:389]: Hm, I am having second thoughts here. Perhaps 

```
pub position: Option<Position>,
pub range: Option<Range>,
```

would be better from the protocol point of view? 

for us, this doesn't really matter, but I wonder if this extension might become used by more then one server. 


Might be a good idea to survery other LSP requests and see what would be the best JSON shape to fit with the rest of the protocol.

Comment (Veykril) [reply crates/rust-analyzer/src/lsp_ext.rs:389]: Not sure if I missed that but wouldn't just a `Range` make more sense where the empty range = position?

Comment (flodiebold) [reply crates/rust-analyzer/src/lsp_ext.rs:389]: Clients still need to be able to pass a `Position` if they don't support the extension.

Comment (Veykril) [reply crates/rust-analyzer/src/lsp_ext.rs:389]: Ah right we need this to be an extension to the original, conveniently forgot that.

Two fields for this feel odd to me, since having both set doesn't make that much sense while its still legal to do? Not sure what the advantages would be of having two fields over one that can be either a range or position.

Comment (matklad) [reply crates/rust-analyzer/src/lsp_ext.rs:389]: to clarify, what specifically feels odd to me is that the field called `position` can actually hold the `Range`.

Comment (alexfertel) [reply crates/rust-analyzer/src/lsp_ext.rs:389]: > to clarify, what specifically feels odd to me is that the field called `position` can actually hold the `Range`. 

That was the first thing that I was blocked by. I think that having separate props and fail if the request is invalid is a bit more clear for someone seeing this for the first time

Comment (matklad) [reply crates/rust-analyzer/src/lsp_ext.rs:389]: Thinking more about this, I think the current solution is fine.

**Comment by simrat39** [comment]
Can the hover response have actions (hover actions) attached to them?

**Comment by Veykril** [comment]
That should be doable just fine since we show a type we can show type hover actions.

**Pull request merged**
